reanalysis/query_panelapp.py

```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python3
+
+
+"""
+PanelApp Parser for Reanalysis
+
+Takes a panel ID
+For the latest content, pulls Symbol, ENSG, and MOI
+    (MOI is simplified from PanelApp enum)
+
+Optionally user can provide a date in the past
+Identify the highest panel version prior to that date
+Pull all details from the earlier version
+Store all discrepancies between earlier and current
+
+Write all output to a JSON dictionary
+"""
+
+
+import logging
+from datetime import datetime
+from typing import Any, Dict, List, Optional, Union
+import json
+
+import requests
+
+import click
+
+
+# panelapp URL constants
+PANELAPP_ROOT = 'https://panelapp.agha.umccr.org/api/v1'
+PANEL_ROOT = f'{PANELAPP_ROOT}/panels'
+PANEL_CONTENT = f'{PANEL_ROOT}/{{panel_id}}'
+ACTIVITIES = f'{PANEL_CONTENT}/activities'
+
+
+def get_json_response(url: str) -> Union[List[Dict[str, str]], Dict[str, Any]]:
+    """
+    takes a request URL, checks for healthy response, returns the JSON
+    :param url:
+    :return:
+    """
+
+    response = requests.get(url, headers={'Accept': 'application/json'})
+    response.raise_for_status()
+    return response.json()
+
+
+def get_previous_version(panel_id: str, since: datetime) -> str:
+    """
+    work through the list of panel updates in reverse (earliest -> latest)
+    return the first panel version after the threshold date
+    If we don't find any, return the latest version
+
+    Note: Django dates include Hours and Minutes, so any Django date from
+    the same day as a 'YYYY-MM-DD' date will be 'greater than'
+    for the purposes of a value comparison
+
+    revisit the exact implementation
+    consider replacement with a panel version (instead of date)
+
+    :param panel_id: panel ID to use
+    :param since: date of the
+    :return:
+    """
+
+    activity_list = get_json_response(url=ACTIVITIES.format(panel_id=panel_id))
+    entry_version = None
+    for entry in reversed(activity_list):
+        # take the panel version
+        entry_version = entry.get('panel_version')
+        # uses the django datestamp format
+        entry_date = datetime.strptime(entry.get('created'), '%Y-%m-%dT%H:%M:%S.%fz')
+
+        if entry_date >= since:
+            return entry_version
+    return entry_version
+
+
+def get_panel_green(
+    reference_genome: Optional[str] = 'GRch38',
+    panel_id: str = '137',
+    version: Optional[str] = None,
+) -> Dict[str, Dict[str, Union[str, bool]]]:
+    """
+    Takes a panel number, and pulls all details from PanelApp
+    :param reference_genome: GRch37 or GRch38
```

Comment:
Is this deliberately spelled with a lower case `c`?

Reply:
Yes 🤦 from PanelApp: `"GRch38": {`


This is irrelevant as the same string is compared in a lowercase form (I expect the capitalisation in PanelApp to be 'corrected' at some point), so I'll make this lower case by default